// react-image-follow/dist/index.js
'use strict';

var React = require('react');
var classNames = require('classnames');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return;
  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

var ImageFollow = function ImageFollow(props) {
  var alt = props.alt,
      className = props.className,
      _props$radius = props.radius,
      radius = _props$radius === void 0 ? 8 : _props$radius,
      src = props.src,
      style = props.style,
      href = props.href,
      children = props.children;

  var _React$useState = React.useState({}),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      cardStyle = _React$useState2[0],
      setCardStyle = _React$useState2[1];

  var _React$useState3 = React.useState({}),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      wrapperStyle = _React$useState4[0],
      setWrapperStyle = _React$useState4[1];

  var _React$useState5 = React.useState({}),
      _React$useState6 = _slicedToArray(_React$useState5, 2),
      imgStyle = _React$useState6[0],
      setImgStyle = _React$useState6[1];

  var cardRef = React.useRef();

  var cardBaseStyle = _objectSpread2({
    borderRadius: "".concat(radius, "px")
  }, style);

  var wrapperBaseStyle = {
    borderRadius: "".concat(radius, "px")
  };
  var imgBaseStyle = {
    borderRadius: "".concat(radius, "px")
  };
  React.useEffect(function () {
    setCardStyle(cardBaseStyle);
    setWrapperStyle(wrapperBaseStyle);
    setImgStyle(imgBaseStyle);
  }, []);
  var cls = classNames__default['default']('react-image-follow', className);

  var move = function move(e) {
    if (!cardRef.current) {
      return;
    }

    var rect = cardRef.current.getBoundingClientRect();
    var _e$nativeEvent = e.nativeEvent,
        clientX = _e$nativeEvent.clientX,
        clientY = _e$nativeEvent.clientY;
    var x = clientX - (rect.left + Math.floor(rect.width / 2));
    var y = clientY - (rect.top + Math.floor(rect.height / 2));
    var transformWrapper = "scale(1.025) translate(calc(".concat(x, " / 50 * 1px), calc(").concat(y, " / 50 * 1px))");
    setWrapperStyle(_objectSpread2(_objectSpread2({}, wrapperBaseStyle), {}, {
      transform: transformWrapper
    }));
    var transformImg = "translate(calc(".concat(x, " / 20 * 1px), calc(").concat(y, " / 20 * 1px))");
    setImgStyle(_objectSpread2(_objectSpread2({}, imgBaseStyle), {}, {
      transform: transformImg
    }));
  };

  var leave = function leave() {
    setWrapperStyle(_objectSpread2(_objectSpread2({}, wrapperBaseStyle), {}, {
      transform: 'none'
    }));
    setImgStyle(_objectSpread2(_objectSpread2({}, imgBaseStyle), {}, {
      transform: 'none'
    }));
  };

  return /*#__PURE__*/React.createElement("a", {
    className: cls,
    style: cardStyle,
    onMouseMove: move,
    onMouseLeave: leave,
    ref: cardRef,
    target: "_blank",
    href: href
  }, /*#__PURE__*/React.createElement("div", {
    style: wrapperStyle,
    className: "react-image-follow-wrapper"
  }, /*#__PURE__*/React.createElement("img", {
    src: src,
    alt: alt,
    className: "react-image-follow-img",
    style: imgStyle
  })), children && /*#__PURE__*/React.createElement("div", {
    className: "react-image-follow-content"
  }, children));
};

module.exports = ImageFollow;
