// react-image-follow/dist/index.esm.js
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import _slicedToArray from '@babel/runtime/helpers/esm/slicedToArray';
import { useState, useRef, useEffect, createElement } from 'react';
import classNames from 'classnames';

var ImageFollow = function ImageFollow(props) {
  var alt = props.alt,
      className = props.className,
      _props$radius = props.radius,
      radius = _props$radius === void 0 ? 8 : _props$radius,
      src = props.src,
      style = props.style,
      href = props.href,
      children = props.children;

  var _React$useState = useState({}),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      cardStyle = _React$useState2[0],
      setCardStyle = _React$useState2[1];

  var _React$useState3 = useState({}),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      wrapperStyle = _React$useState4[0],
      setWrapperStyle = _React$useState4[1];

  var _React$useState5 = useState({}),
      _React$useState6 = _slicedToArray(_React$useState5, 2),
      imgStyle = _React$useState6[0],
      setImgStyle = _React$useState6[1];

  var cardRef = useRef();

  var cardBaseStyle = _objectSpread({
    borderRadius: "".concat(radius, "px")
  }, style);

  var wrapperBaseStyle = {
    borderRadius: "".concat(radius, "px")
  };
  var imgBaseStyle = {
    borderRadius: "".concat(radius, "px")
  };
  useEffect(function () {
    setCardStyle(cardBaseStyle);
    setWrapperStyle(wrapperBaseStyle);
    setImgStyle(imgBaseStyle);
  }, []);
  var cls = classNames('react-image-follow', className);

  var move = function move(e) {
    if (!cardRef.current) {
      return;
    }

    var rect = cardRef.current.getBoundingClientRect();
    var _e$nativeEvent = e.nativeEvent,
        clientX = _e$nativeEvent.clientX,
        clientY = _e$nativeEvent.clientY;
    var x = clientX - (rect.left + Math.floor(rect.width / 2));
    var y = clientY - (rect.top + Math.floor(rect.height / 2));
    var transformWrapper = "scale(1.025) translate(calc(".concat(x, " / 50 * 1px), calc(").concat(y, " / 50 * 1px))");
    setWrapperStyle(_objectSpread(_objectSpread({}, wrapperBaseStyle), {}, {
      transform: transformWrapper
    }));
    var transformImg = "translate(calc(".concat(x, " / 20 * 1px), calc(").concat(y, " / 20 * 1px))");
    setImgStyle(_objectSpread(_objectSpread({}, imgBaseStyle), {}, {
      transform: transformImg
    }));
  };

  var leave = function leave() {
    setWrapperStyle(_objectSpread(_objectSpread({}, wrapperBaseStyle), {}, {
      transform: 'none'
    }));
    setImgStyle(_objectSpread(_objectSpread({}, imgBaseStyle), {}, {
      transform: 'none'
    }));
  };

  return /*#__PURE__*/createElement("a", {
    className: cls,
    style: cardStyle,
    onMouseMove: move,
    onMouseLeave: leave,
    ref: cardRef,
    target: "_blank",
    href: href
  }, /*#__PURE__*/createElement("div", {
    style: wrapperStyle,
    className: "react-image-follow-wrapper"
  }, /*#__PURE__*/createElement("img", {
    src: src,
    alt: alt,
    className: "react-image-follow-img",
    style: imgStyle
  })), children && /*#__PURE__*/createElement("div", {
    className: "react-image-follow-content"
  }, children));
};

export default ImageFollow;
